feat(parentheticals): Add support for scraping and storing citation parentheticals

ProbablyFaiz opened the pull request:
This PR adds support for scraping and storing [citation parentheticals](https://www.law.georgetown.edu/wp-content/uploads/2018/07/Parentheticals-Bluebook-Handout-Revision-Karl-Bock-2016.pdf) to CourtListener. Specifically, it does the following:

- Adds a new `OpinionDescription` table to store parentheticals. (It's named with "description" rather than "parenthetical" to allow for the future possibility of descriptions scraped from sources other than post-citation parentheticals)
- Adds a Regex ruleset and helper methods for filtering out parentheticals that aren't descriptive (e.g. "Scalia, J., concurring").
- Modifies the `cl_find_citations` command to scrape and store parenthetical descriptions as well as citations found in opinion text.

Currently, these parentheticals are not exposed anywhere on the frontend. But this PR lays the requisite groundwork for those changes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/freelawproject/courtlistener?pullRequest=1903) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/freelawproject/courtlistener?pullRequest=1903) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/freelawproject/courtlistener?pullRequest=1903) it.</sub>

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+
+_SURROUNDING_CHARS = r'[.!;,"“” ]'
+
+PARENTHETICAL_REGEX_BLACKLIST_RULES = [
```
I try not to use terms like "blacklist" anymore. 

May I suggest:

 - BLOCKLIST
 - NOLIST
 - BADLIST

Or a different variable name you prefer better?

mlissner commented on cl/citations/tests.py:
```diff
@@ -761,3 +792,113 @@ def test_making_edge_list(self) -> None:
+            ("collecting cases", False),
+        ]
+
+        for parenthetical_text, expected_is_descriptive in test_pairs:
+            self.assertEqual(
+                is_parenthetical_descriptive(parenthetical_text),
+                expected_is_descriptive,
+                f"Got incorrect result from is_parenthetical_descriptive for text (expected {expected_is_descriptive}): {parenthetical_text}",
+            )
```
1. This and your following test would benefit from the use of self.subTest. 
2. I'd split up this test into two: test_is_good_description and test_not_not_good_description (or similar). That'd add some organization to your True/False test cases beyond putting the Falses latter, and it'd simplify your test variable so that it didn't have nested tuples.

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+    r".{1,25} I+",  # Gonzales II
+    r".{1,10} (?:Circuit|Cir.)",  # Tenth Circuit, 5th Cir.
+    r"here(?:in)?after .+",  # hereinafter, "Griffin II"
+    r"\S*",  # Single-word parentheticals, e.g., 'TILA'
```
This, and perhaps some others, seem like they could be eliminated by adding a length requirement. Would it be helpful to require at least so many characters?

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+_HONORIFICS = r"(?:Mr.?|Mister)"
+_JUDGE_NAME = rf"(?:(?:.{{1,25}}J\.,?)|(?:{_HONORIFICS} Justice .{{1,25}})|the Court|(?:.{{1,25}}Circuit Justice),?)"
+
+_SURROUNDING_CHARS = r'[.!;,"“” ]'
```
This would be better placed just about prefix and suffix.

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+import re
+
+_MODIFIABLE = r"(?:omissions?|quotations?|quotes?|headings?|(?:quotations? )?marks?|ellips.s|cites?|citations?|emphas.s|italics?|footnotes?|alterations?|punctuation|modifications?|brackets?|bracketed material|formatting)"
+_MODIFABLE_TYPE = r"(?:internal|former|latter|first|second|last|some|further|certain|numbered|other|transcript)"
```
Worth adding third, fouth, fifth?

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+import re
```
Two brought thoughts on this file:

1. Regexes are just the worst. Too bad they're also the best. Given that, would it be possible to take a run at refactoring this file to make it even cleaner than it is already? I've highlighted a couple things so far.

2. I think another thing that'd help — I think — would be to eliminate the non-capturing groups. Unless I'm mistaken capturing or not is fine? That'd eliminate a bunch of question marks.

mlissner replied:
Sorry, one more: 

3. Are all of these tested? I see a bunch of tests, but haven't checked if everything here has a test. With something like this, a big test suite is going to be essential.

ProbablyFaiz replied:
1. https://xkcd.com/1171/ Yep, I'll make those changes and see what else I can do to clean things up.
2. That's a good point, I'll do that.
3. Many but definitely not all of the different cases are tested. We will add some more test cases to make it a bit more comprehensive.

mlissner commented on cl/citations/tests.py:
```diff
@@ -761,3 +792,113 @@ def test_making_edge_list(self) -> None:
+                "ruling that there is nothing either legal or illegal, only thinking makes it so",
+                True,
+            ),
+            ("First Amendment", True),
+            ("mislabeled product", True),
+            ('"Look on my Works, ye Mighty, and despair"', True),
+            (
+                '"Texas does not seek to have the Court interpret the Constitution, so much as disregard it."',
+                True,
+            ),
+            ("questioning whether he who made the Lamb made thee", True),
+            ("Section 403(d)(2)", True),
```
All of these register to me as ones that should be False, right?

ProbablyFaiz replied:
This gets at a philosophical decision as to what we consider useful. 

The approach we took was to try to filter out what we considered citation "metadata" (or "spam", when we talked about it) while preserving as much as possible anything "substantive". While something like "First Amendment" or "mislabeled product" is not very useful to show on the opinion page on CourtListener (and would be downranked by our heuristic in #1906), we thought it could still potentially be of interest to a researcher were these parentheticals to be published in some kind of bulk dataset.

That's the reason -- to address a couple of your other comments -- we took a pretty fine-grained approach to filtering and didn't place a blanket length restriction. I can definitely see though how this increases complexity and maintenance burden, though, so I'm open to whatever you think is the best course of action.

mlissner replied:
Gotcha. I'll add another argument to your side: If we gather everything in the DB, we can more easily see what we have and refine from there. You've convinced me here.

mlissner commented on cl/citations/filter_parentheticals.py:
```diff
@@ -0,0 +1,76 @@
+    return not re.match(PARENTHETICAL_BLACKLIST_REGEX, text)
+
+
+def clean_parenthetical_text(text: str) -> str:
+    text = re.sub(
+        r"\* ?\d+", " ", text
+    )  # Remove star page numbers (e.g. *389)
+    text = re.sub(
+        r" [_-]{4,} ", " ", text
+    )  # Remove weird "word ----- word" stuff (but don't screw up something like "123 U.S. ---")
+    text = re.sub(
+        r"\s+", " ", text
+    )  # Remove excess whitespace that may have appeared as a result of sequential dash sequences
+    return text
```
Just a style thing, and I know you've satisfied `black`, but I think I'd rather see these comments on their own lines above what they do, rather than trailing after. Don't feel obligated to fix this, but this looks weird to me.

mlissner commented on cl/citations/tasks.py:
```diff
@@ -148,6 +160,23 @@ def find_citations_for_opinion_by_pks(
+                    # If the citation has a descriptive parenthetical, clean
+                    # it up and store it as an OpinionDescription
+                    if (
+                        par_text := _cit.metadata.parenthetical
```
Congrats! This will land the first walrus operator into the codebase.

mlissner reviewed (changes requested):
OK, this is very nice all in all, thank you! Nice job getting this added so quickly. 

A couple high-level things:

1. I'm unconvinced that calling this feature descriptions is a good idea. I see the intent: We might add something different down the road and so it's good to have a broad table name now that will support new features later. That makes sense, particularly since it's basically impossible to rename tables down the road. 

    However. Given the *shape* of the object, and that it links up two opinions with some metadata (the text), I can't think of another kind of description that this model would encompass down the road. I.e., I can't think of another type of description from one opinion that's used to describe another opinion.

    I'm happy to hear dissent here, but if that's the case, we should rename this to be more accurate since that will make the code more understandable. Perhaps just `Parenthetical`, for example, would be nice?

1. Can you please rename your migration to 0005_add_opinion_parentheticals (or "_descriptions", if you convince me description is better)?

3.We need SQL migrations for any DB migrations. If you run manage.py sqlmigrate search 0005, it'll print out some SQL that you can throw in a migration file. Poke around in the migration directories and you'll see an example of one of these.

That's it for now. It's looking very good.

ProbablyFaiz commented:
> 1. I'm unconvinced that calling this feature descriptions is a good idea. I see the intent: We might add something different down the road and so it's good to have a broad table name now that will support new features later. That makes sense, particularly since it's basically impossible to rename tables down the road.
>    However. Given the _shape_ of the object, and that it links up two opinions with some metadata (the text), I can't think of another kind of description that this model would encompass down the road. I.e., I can't think of another type of description from one opinion that's used to describe another opinion.
>    I'm happy to hear dissent here, but if that's the case, we should rename this to be more accurate since that will make the code more understandable. Perhaps just `Parenthetical`, for example, would be nice?

My best shot is that Casetext's summaries feature, while mostly sourced from parentheticals, does seem to pull in some other patterns too. Here's one I found on a random case on Casetext:
> In Mnemonics, we applied this test to Florida trade secret law, and recognized that the breach of duty essential to a trade secret case is an extra element that qualitatively distinguishes those claims from copyright claims.

Looking for sentences like that is probably more trouble than it's worth. But given that it is *possible* to do so, I made the name general. I'm certainly not wedded to it though and we can change that if you're not sold.

mlissner commented:
That's interesting, but I think I'm not sold, yeah. It's a bit too much of a long shot for it happening and in the meantime the code is less friendly to new folks. Maybe if I saw it in practice, if you had a link to the above, @varun-iyer ? But I don't think I'm feeling it so far.

ProbablyFaiz commented:
Alright, I think we've addressed the issues here and this PR is ready for another look.

- Renamed the table and associated references to `Parenthetical`. Renamed the migration file and added the corresponding SQL
- Cleaned up the regex file a bit and added more comprehensive tests
- Merged the migration from #1906 into this one.

mlissner commented on cl/search/models.py:
```diff
@@ -2846,6 +2846,30 @@ class Meta:
+    text = models.TextField(
+        help_text="The text of the description as written in the describing opinion",
+    )
+    usefulness_score = models.FloatField(
+        db_index=True,
+        help_text="A score between 0 and 1 representing how descriptive the parenthetical is",
+    )
```
We'll need this to have a default or be nullable, won't we?

mlissner replied:
I think I would also just call this `score`, to keep it simple. That matches nicely with the simplicity of `text` too.

mlissner replied:
Just need this to have a default or be nullable and I think we're there.

ProbablyFaiz replied:
Done

mlissner commented:
One small comment from a second pass, above.

The other thing is that @flooie has been watching this and is advocating for a different name for the table or maybe the feature, but I haven't had a chance to talk to him yet. Hopefully he I'll convince him it's fine as it is, but he's often right when he gets to thinking about stuff. I'll report back from that as soon as he and I can talk.

mlissner commented:
I might be mistaken, but it looks like the sql file still needs an update?

varun-iyer commented:
> I might be mistaken, but it looks like the sql file still needs an update?

Hmm, it looks to me like the SQL output of `manage.py sqlmigrate search 0005` is identical to the .sql file in the latest commit here ([b86ca36](https://github.com/freelawproject/courtlistener/commit/b86ca36d00468ad1baa6a6c66752f83a77d3dd50) in case there's a force push).

However, I've never generated SQL migration files in this way before --- is there a step I'm missing here or am I misunderstanding the generation process?

mlissner commented:
> Apparently Django’s ORM handles default values exclusively at the *application* level, rather than the database level:

You know, this rings a bell. I was getting confused because other migrations *do* set default values temporarily when you have `blank=True`. When that happens, it causes slightly older versions of postgres to lose their mind: https://github.com/freelawproject/courtlistener/blob/main/MIGRATIONS.md#new-columns-with-default-values-in-postgresql--v11

So I was mixing up the memory of that to think that there ought to be a default *something* in the SQL, and I was forgetting that defaults are handled applicationside. Great.

Thank you.

mlissner commented:
OK, that's one! Thank you guys!
